### PR TITLE
Implement ICS 04

### PR DESCRIPTION
**Pull request by junha1**
There are still missing required functions, since we don't need it in PoC

### CODE REVIEW

File: core/src/ibc/channel_04/log.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/// Nasty version of blockchain log event system.
+/// In PoC, we consider only ORDERED channel.
+/// Thus, it is enough to store only latest packet per channel.
+/// Since we have no log system yet, we just store it in state DB.
+///
+use crate::ibc;
+use ccrypto::blake256;
+use ibc::IdentifierSlice;
+
+pub fn set_packet<'a>(ctx: &'a mut dyn ibc::Context, packet: &Packet, tag: &str) {
```
Comment (junha1) [inline core/src/ibc/channel_04/log.rs]: Here `tag` is either `send` or `recv`

File: core/src/ibc/channel_04/mod.rs [hunk 2]
```diff
@@ -14,10 +14,10 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+mod log;
 mod manager;
 pub mod types;
 
-
```
Comment (majecty) [inline core/src/ibc/channel_04/mod.rs]: Could you create another commit for this change?

File: core/src/ibc/channel_04/manager.rs [hunk 3]
```diff
@@ -409,4 +413,60 @@ impl<'a> Manager<'a> {
         self.ctx.get_kv_store_mut().insert(&channel_path(DEFAULT_PORT, &channel_identifier), &rlp::encode(&channel));
         Ok(())
     }
+
+    //
+    // PACKET
+    //
```
Comment (majecty) [inline core/src/ibc/channel_04/manager.rs]: What is the purpose of this comment?

File: core/src/ibc/channel_04/manager.rs [hunk 4]
```diff
@@ -469,4 +470,73 @@ impl<'a> Manager<'a> {
...
+        if packet.source_channel != channel.counterparty_channel_identifier {
+            return Err("Packet's source_channel doesn't match with channel.".to_owned())
+        }
+
+        self.check_capability_key(&packet.dest_port, &packet.dest_channel)?;
+        let client_identifier = self.check_connection_opened(&channel.connection_hops[0])?;
+        let client_manager = ClientManager::new(self.ctx);
+        let client_state = client_manager.query(&client_identifier)?;
+
+
```
Comment (majecty) [inline core/src/ibc/channel_04/manager.rs]: Let's remove this line.

File: core/src/ibc/channel_04/mod.rs (hunk 2 above)
Comment (junha1) [reply core/src/ibc/channel_04/mod.rs]: done

**Pull request merged**